File: backend/app/schemas.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field, model_validator


class AuthRequest(BaseModel):
    email: EmailStr = Field(..., description="Coder account email")
    password: str = Field(..., min_length=1, description="Coder account password")


class AuthResponse(BaseModel):
    session_token: str = Field(
        ...,
        description="Coder session token. Send as Coder-Session-Token on subsequent API calls.",
    )


class MeResponse(BaseModel):
    username: str = Field(..., description="Coder username")
    email: str | None = None
    dashboard_url: str = Field(
        ...,
        description="Coder dashboard base URL used for web app links.",
    )


class WorkspaceAccessResponse(BaseModel):
    workspace_id: UUID
    workspace_name: str
    username: str
    started: bool
    startup_ready: bool = False
    agent_lifecycle_state: str | None = None
    agent_id: UUID | None = None
    agent_name: str | None = None
    agent_directory: str | None = None
    has_terminal: bool = False
    has_vscode_browser: bool = False
    has_vscode_desktop: bool = False
    vscode_browser_url: str | None = None
    code_server_slug: str | None = None


class VSCodeDesktopResponse(BaseModel):
    uri: str = Field(..., description="vscode://coder.coder-remote/open?... deep link")


class RichParameterValue(BaseModel):
    name: str = Field(..., min_length=1, description="Parameter name (e.g. cpu, memory)")
    value: str = Field(..., min_length=1, description="Parameter value as a string")


class CreateWorkspaceRequest(BaseModel):
    name: str = Field(
        ...,
        min_length=1,
        max_length=32,
        pattern=r"^[a-zA-Z0-9]([a-zA-Z0-9-]{0,31})$",
        description="Workspace name (letters, numbers, hyphens; max 32 chars).",
    )
    template_id: UUID | None = Field(
        default=None,
        description="Template ID (uses active version). Mutually exclusive with template_version_id.",
    )
    template_version_id: UUID | None = Field(
        default=None,
        description="Specific template version. Mutually exclusive with template_id.",
    )
    rich_parameter_values: list[RichParameterValue] = Field(
        default_factory=list,
        description="Rich parameter values for the initial provision.",
    )

    @model_validator(mode="after")
    def validate_template_source(self) -> CreateWorkspaceRequest:
        if self.name.lower() in {"new", "create"}:
            raise ValueError("Workspace name cannot be 'new' or 'create'")
        has_template = self.template_id is not None
        has_version = self.template_version_id is not None
        if has_template == has_version:
            raise ValueError("Provide exactly one of template_id or template_version_id")
        return self

    def to_coder_body(self) -> dict[str, Any]:
        body: dict[str, Any] = {"name": self.name}
        if self.template_id is not None:
            body["template_id"] = str(self.template_id)
        if self.template_version_id is not None:
            body["template_version_id"] = str(self.template_version_id)
        if self.rich_parameter_values:
            body["rich_parameter_values"] = [
                {"name": p.name, "value": p.value} for p in self.rich_parameter_values
            ]
        return body


class BuildSummary(BaseModel):
    id: UUID
    status: str = Field(..., description="Job status: pending | running | succeeded | failed | canceled")
    build_number: int | None = None
    transition: str | None = None


class WorkspaceResponse(BaseModel):
    id: UUID
    name: str
    template_id: UUID | None = None
    latest_build: BuildSummary
    agent_lifecycle_state: str | None = None
    startup_ready: bool = False

    @classmethod
    def from_coder(cls, payload: dict[str, Any]) -> WorkspaceResponse:
        from app.workspace_access import agent_lifecycle_state, is_agent_ready, pick_agent

        latest = payload.get("latest_build") or {}
        job = latest.get("job") or {}
        agent = pick_agent(payload)
        return cls(
            id=payload["id"],
            name=payload["name"],
            template_id=payload.get("template_id"),
            latest_build=BuildSummary(
                id=latest["id"],
                status=job.get("status") or "unknown",
                build_number=latest.get("build_number"),
                transition=latest.get("transition"),
            ),
            agent_lifecycle_state=agent_lifecycle_state(agent),
            startup_ready=is_agent_ready(agent),
        )


class WorkspaceListResponse(BaseModel):
    count: int
    workspaces: list[WorkspaceResponse]


class WorkspaceBuildResponse(BaseModel):
    id: UUID
    workspace_id: UUID
    build_number: int
    status: str = Field(..., description="Job status: pending | running | succeeded | failed | canceled")
    transition: str | None = None
    job_error: str | None = None
    created_at: datetime | None = None

    @classmethod
    def from_coder(cls, payload: dict[str, Any]) -> WorkspaceBuildResponse:
        job = payload.get("job") or {}
        return cls(
            id=payload["id"],
            workspace_id=payload["workspace_id"],
            build_number=payload["build_number"],
            status=job.get("status") or "unknown",
            transition=payload.get("transition"),
            job_error=job.get("error") or None,
            created_at=payload.get("created_at"),
        )


class ProvisionerJobLog(BaseModel):
    id: int
    created_at: datetime | None = None
    log_level: str | None = None
    log_source: str | None = None
    output: str | None = None
    stage: str | None = None


class WorkspaceAgentLog(BaseModel):
    id: int
    created_at: datetime | None = None
    level: str | None = None
    output: str | None = None
    source_id: UUID | None = None


LogFormat = Literal["json", "text"]

File: go-coder-service/internal/api/schemas.go
// Package api exposes the JSON API, mirroring backend/app/{schemas,main}.py.
package api

import (
	"fmt"
	"net/http"
	"regexp"
	"strings"

	"github.com/google/uuid"

	"github.com/malmonte/go-coder-service/internal/coder"
	"github.com/malmonte/go-coder-service/internal/config"
)

type AuthRequest struct {
	Email string `json:"email"`
}

type AuthResponse struct {
	SessionToken string `json:"session_token"`
}

type MeResponse struct {
	Username     string  `json:"username"`
	Email        *string `json:"email"`
	DashboardURL string  `json:"dashboard_url"`
}

type WorkspaceAccessResponse struct {
	WorkspaceID         string  `json:"workspace_id"`
	WorkspaceName       string  `json:"workspace_name"`
	Username            string  `json:"username"`
	Started             bool    `json:"started"`
	StartupReady        bool    `json:"startup_ready"`
	AgentLifecycleState *string `json:"agent_lifecycle_state"`
	AgentID             *string `json:"agent_id"`
	AgentName           *string `json:"agent_name"`
	AgentDirectory      *string `json:"agent_directory"`
	HasTerminal         bool    `json:"has_terminal"`
	TerminalURL         *string `json:"terminal_url"`
	HasVSCodeBrowser    bool    `json:"has_vscode_browser"`
	HasVSCodeDesktop    bool    `json:"has_vscode_desktop"`
	VSCodeBrowserURL    *string `json:"vscode_browser_url"`
	CodeServerSlug      *string `json:"code_server_slug"`
}

type VSCodeDesktopResponse struct {
	URI string `json:"uri"`
}

type RichParameterValue struct {
	Name  string `json:"name"`
	Value string `json:"value"`
}

type CreateWorkspaceRequest struct {
	Name                string               `json:"name"`
	TemplateID          *string              `json:"template_id"`
	TemplateVersionID   *string              `json:"template_version_id"`
	RichParameterValues []RichParameterValue `json:"rich_parameter_values"`
}

var workspaceNamePattern = regexp.MustCompile(`^[a-zA-Z0-9]([a-zA-Z0-9-]{0,31})$`)

// Validate replicates the pydantic model constraints on CreateWorkspaceRequest.
func (r *CreateWorkspaceRequest) Validate() error {
	if len(r.Name) < 1 || len(r.Name) > 32 || !workspaceNamePattern.MatchString(r.Name) {
		return coder.NewAPIError(http.StatusUnprocessableEntity,
			"Workspace name must be letters, numbers, or hyphens (max 32 chars)")
	}
	lower := strings.ToLower(r.Name)
	if lower == "new" || lower == "create" {
		return coder.NewAPIError(http.StatusUnprocessableEntity,
			"Workspace name cannot be 'new' or 'create'")
	}
	hasTemplate := r.TemplateID != nil && *r.TemplateID != ""
	hasVersion := r.TemplateVersionID != nil && *r.TemplateVersionID != ""
	if hasTemplate == hasVersion {
		return coder.NewAPIError(http.StatusUnprocessableEntity,
			"Provide exactly one of template_id or template_version_id")
	}
	for _, id := range []*string{r.TemplateID, r.TemplateVersionID} {
		if id != nil && *id != "" {
			if _, err := uuid.Parse(*id); err != nil {
				return coder.NewAPIError(http.StatusUnprocessableEntity,
					fmt.Sprintf("Invalid UUID: %s", *id))
			}
		}
	}
	for _, param := range r.RichParameterValues {
		if param.Name == "" || param.Value == "" {
			return coder.NewAPIError(http.StatusUnprocessableEntity,
				"Rich parameter values require non-empty name and value")
		}
	}
	return nil
}

func (r *CreateWorkspaceRequest) ToCoderBody() map[string]any {
	body := map[string]any{"name": r.Name}
	if r.TemplateID != nil && *r.TemplateID != "" {
		body["template_id"] = *r.TemplateID
	}
	if r.TemplateVersionID != nil && *r.TemplateVersionID != "" {
		body["template_version_id"] = *r.TemplateVersionID
	}
	if len(r.RichParameterValues) > 0 {
		params := make([]map[string]any, 0, len(r.RichParameterValues))
		for _, p := range r.RichParameterValues {
			params = append(params, map[string]any{"name": p.Name, "value": p.Value})
		}
		body["rich_parameter_values"] = params
	}
	return body
}

type BuildSummary struct {
	ID          string  `json:"id"`
	Status      string  `json:"status"`
	BuildNumber *int64  `json:"build_number"`
	Transition  *string `json:"transition"`
}

type WorkspaceResponse struct {
	ID                  string                   `json:"id"`
	Name                string                   `json:"name"`
	TemplateID          *string                  `json:"template_id"`
	LatestBuild         BuildSummary             `json:"latest_build"`
	AgentLifecycleState *string                  `json:"agent_lifecycle_state"`
	StartupReady        bool                     `json:"startup_ready"`
	Access              *WorkspaceAccessResponse `json:"access"`
}

type WorkspaceListResponse struct {
	Count      int64               `json:"count"`
	Workspaces []WorkspaceResponse `json:"workspaces"`
}

type WorkspaceBuildResponse struct {
	ID          string  `json:"id"`
	WorkspaceID string  `json:"workspace_id"`
	BuildNumber int64   `json:"build_number"`
	Status      string  `json:"status"`
	Transition  *string `json:"transition"`
	JobError    *string `json:"job_error"`
	CreatedAt   *string `json:"created_at"`
}

type ProvisionerJobLog struct {
	ID        int64   `json:"id"`
	CreatedAt *string `json:"created_at"`
	LogLevel  *string `json:"log_level"`
	LogSource *string `json:"log_source"`
	Output    *string `json:"output"`
	Stage     *string `json:"stage"`
}

type WorkspaceAgentLog struct {
	ID        int64   `json:"id"`
	CreatedAt *string `json:"created_at"`
	Level     *string `json:"level"`
	Output    *string `json:"output"`
	SourceID  *string `json:"source_id"`
}

func optString(obj map[string]any, key string) *string {
	if obj == nil {
		return nil
	}
	if value, ok := obj[key].(string); ok && value != "" {
		return &value
	}
	return nil
}

func optInt(obj map[string]any, key string) *int64 {
	if obj == nil {
		return nil
	}
	if value, ok := obj[key].(float64); ok {
		i := int64(value)
		return &i
	}
	return nil
}

func intOr(obj map[string]any, key string, fallback int64) int64 {
	if value := optInt(obj, key); value != nil {
		return *value
	}
	return fallback
}

func str(obj map[string]any, key string) string {
	if obj == nil {
		return ""
	}
	value, _ := obj[key].(string)
	return value
}

func obj(parent map[string]any, key string) map[string]any {
	if parent == nil {
		return nil
	}
	value, _ := parent[key].(map[string]any)
	return value
}

// WorkspaceResponseFromCoder mirrors WorkspaceResponse.from_coder.
func WorkspaceResponseFromCoder(payload map[string]any) WorkspaceResponse {
	latest := obj(payload, "latest_build")
	job := obj(latest, "job")
	agent := coder.PickAgent(payload)
	status := str(job, "status")
	if status == "" {
		status = "unknown"
	}
	return WorkspaceResponse{
		ID:         str(payload, "id"),
		Name:       str(payload, "name"),
		TemplateID: optString(payload, "template_id"),
		LatestBuild: BuildSummary{
			ID:          str(latest, "id"),
			Status:      status,
			BuildNumber: optInt(latest, "build_number"),
			Transition:  optString(latest, "transition"),
		},
		AgentLifecycleState: optString(agent, "lifecycle_state"),
		StartupReady:        coder.IsAgentReady(agent),
	}
}

// WorkspaceBuildResponseFromCoder mirrors WorkspaceBuildResponse.from_coder.
func WorkspaceBuildResponseFromCoder(payload map[string]any) WorkspaceBuildResponse {
	job := obj(payload, "job")
	status := str(job, "status")
	if status == "" {
		status = "unknown"
	}
	return WorkspaceBuildResponse{
		ID:          str(payload, "id"),
		WorkspaceID: str(payload, "workspace_id"),
		BuildNumber: intOr(payload, "build_number", 0),
		Status:      status,
		Transition:  optString(payload, "transition"),
		JobError:    optString(obj(payload, "job"), "error"),
		CreatedAt:   optString(payload, "created_at"),
	}
}

func ProvisionerJobLogFromCoder(entry map[string]any) ProvisionerJobLog {
	return ProvisionerJobLog{
		ID:        intOr(entry, "id", 0),
		CreatedAt: optString(entry, "created_at"),
		LogLevel:  optString(entry, "log_level"),
		LogSource: optString(entry, "log_source"),
		Output:    optString(entry, "output"),
		Stage:     optString(entry, "stage"),
	}
}

func WorkspaceAgentLogFromCoder(entry map[string]any) WorkspaceAgentLog {
	return WorkspaceAgentLog{
		ID:        intOr(entry, "id", 0),
		CreatedAt: optString(entry, "created_at"),
		Level:     optString(entry, "level"),
		Output:    optString(entry, "output"),
		SourceID:  optString(entry, "source_id"),
	}
}

// WorkspaceAccess mirrors main.py _workspace_access.
func WorkspaceAccess(
	workspace, user map[string]any, dashboard string, settings *config.Settings,
) WorkspaceAccessResponse {
	name := str(workspace, "name")
	appBase := dashboard
	if !settings.UsesSSOAccess() {
		appBase = coder.PreferCoderAppBase(settings.CoderAPIBase(), dashboard)
	}
	username := str(user, "username")
	started := coder.IsStarted(workspace)
	agent := coder.PickAgent(workspace)
	ready := started && coder.IsAgentReady(agent)

	access := WorkspaceAccessResponse{
		WorkspaceID:         str(workspace, "id"),
		WorkspaceName:       name,
		Username:            username,
		Started:             started,
		StartupReady:        ready,
		AgentLifecycleState: optString(agent, "lifecycle_state"),
		AgentID:             optString(agent, "id"),
		AgentName:           optString(agent, "name"),
	}
	if agent != nil {
		if directory := str(agent, "expanded_directory"); directory != "" {
			access.AgentDirectory = &directory
		} else if directory := str(agent, "directory"); directory != "" {
			access.AgentDirectory = &directory
		}
	}
	codeApp := coder.FindApp(agent, "code-server")
	apps := coder.DisplayApps(agent)
	hasVSCodeDesktop := false
	for _, app := range apps {
		if app == "vscode" || app == "vscode_insiders" {
			hasVSCodeDesktop = true
			break
		}
	}
	if ready && access.AgentName != nil && codeApp != nil {
		slug := str(codeApp, "slug")
		if slug == "" {
			slug = "code-server"
		}
		url := coder.BuildVSCodeBrowserURL(appBase, username, name, *access.AgentName, slug)
		access.VSCodeBrowserURL = &url
	}
	if ready && access.AgentID != nil && settings.UsesSSOAccess() {
		agentName := ""
		if access.AgentName != nil {
			agentName = *access.AgentName
		}
		url := coder.BuildWebTerminalURL(dashboard, username, name, agentName)
		access.TerminalURL = &url
	}
	access.HasTerminal = ready && access.AgentID != nil
	access.HasVSCodeBrowser = ready && codeApp != nil
	access.HasVSCodeDesktop = ready && hasVSCodeDesktop
	if codeApp != nil {
		access.CodeServerSlug = optString(codeApp, "slug")
	}
	return access
}
